pkg/controller/authentication/service.go

```diff
@@ -46,6 +46,33 @@ func (r *ReconcileAuthentication) handleService(instance *operatorv1alpha1.Authe
 	} else if err != nil {
 		reqLogger.Error(err, "Failed to get Service")
 		return err
+	} else {
```

Comment:
It looks as if the following code could/should be broken out into its own function and called once for each of these services. Something like this maybe?

```golang
// validateCP3ServicePodSelectorAndLabel takes a *Service and attempts to update that Service's selectors and
// labels if they do not match the desired CP3 values. Returns an error if the update fails or if the provided
// *Service is nil or lacks a name or namespace.
func validateCP3PodSelectorAndLabel(currentService *corev1.Service) (err error) {
	if currentService == nil || currentService.Name == "" || currentService.Namespace == "" {
		return fmt.Errorf("received invalid Service")
	}
	var cp2podselector bool = false
	var cp2podlabel bool = false
	podSelector := currentService.Spec.Selector
	value, ok := podSelector["k8s-app"]
	if ok && value != currentService.Name {
		currentService.Spec.Selector = map[string]string{"k8s-app": currentService.Name}
		cp2podselector = true
	}
	// Going to validate label for CP3 upgrade
	label := currentService.Labels
	value, ok = label["app"]
	if ok && value != currentService.Name {
		currentService.Labels = map[string]string{"app": currentService.Name}
		cp2podlabel = true
	}
	if cp2podselector || cp2podlabel {
		err = r.client.Update(context.Background(), currentService)
		if err != nil {
			reqLogger.Error(err, "Upgrade check : Failed to update service podSelector , label details ", "Service.Namespace", instance.Namespace, "Service.Name", currentService.Name, "Error.message", err)
			r.needToRequeue = true
			return
		} else {
			reqLogger.Info("Upgrade check : Successfully updated service podSelector , label details ", "Service.Name", currentService.Name)
		}
	}
	return
}

Reply:
Thanks for your suggestion , I changed accordingly.